internal/extender/resourcereservations.go

```diff
@@ -15,26 +15,281 @@
+	resourceReservation, ok := rrm.GetResourceReservation(executor)
+	if !ok {
+		return "", werror.ErrorWithContextParams(ctx, "failed to get resource reservations")
+	}
+	for name := range resourceReservation.Spec.Reservations {
+		if resourceReservation.Status.Pods[name] == executor.Name {
+			return resourceReservation.Spec.Reservations[name].Node, nil
+		}
+	}
+
+	// Check any soft reservations as well
+	if sr, ok := rrm.softReservationStore.GetExecutorSoftReservation(ctx, executor); ok {
+		return sr.Node, nil
+	}
+	return "", nil
+}
+
+// FindUnboundReservationNodes returns a slice of node names that have unbound reservations for this Spark application.
+// This includes both reservations we have not yet scheduled any executors on as well as reservations that have executors that are now dead.
+// Spark will recreate lost executors, so the replacement executors should be placed on the reserved spaces of dead executors.
+func (rrm *ResourceReservationManager) FindUnboundReservationNodes(ctx context.Context, executor *v1.Pod) ([]string, error) {
+	unboundReservationsToNodes, err := rrm.getUnboundReservations(ctx, executor)
+	if err != nil {
+		return []string{}, err
+	}
+	unboundReservationNodes := utils.NewStringSet(len(unboundReservationsToNodes))
+	for _, node := range unboundReservationsToNodes {
+		unboundReservationNodes.Add(node)
+	}
+	return unboundReservationNodes.ToSlice(), nil
+}
+
+// GetFreeExecutorSpots returns the number of executors the application can still schedule.
+func (rrm *ResourceReservationManager) GetFreeExecutorSpots(ctx context.Context, executor *v1.Pod) (int, error) {
+	unboundReservations, err := rrm.getUnboundReservations(ctx, executor)
+	if err != nil {
+		return 0, err
+	}
+	extraExecutorFreeSpots, err := rrm.getFreeExtraExecutorSpots(ctx, executor)
+	if err != nil {
+		return 0, err
+	}
+	return len(unboundReservations) + extraExecutorFreeSpots, nil
+}
+
+// ReserveForExecutor creates a reservation for the passed executor on the passed node. This reservation could either be a resource reservation,
+// or a soft reservation if dynamic allocation is enabled.
+func (rrm *ResourceReservationManager) ReserveForExecutor(ctx context.Context, executor *v1.Pod, node string) error {
+	rrm.mutex.Lock()
+	defer rrm.mutex.Unlock()
+
+	// TODO: make sure executor doesn't already have a reservation, if it does, it must be freed (we don't seem to handle this in the current logic, but we should going forward)
+
+	unboundReservationsToNodes, err := rrm.getUnboundReservations(ctx, executor)
+	if err != nil {
+		return err
+	}
+	// Sort the unbound reservations such that we favor reservations that are already tied to the requested node first
+	sortedUnboundReservations := make([]string, len(unboundReservationsToNodes))
+	for reservation := range unboundReservationsToNodes {
+		sortedUnboundReservations = append(sortedUnboundReservations, reservation)
+	}
+	sort.Slice(sortedUnboundReservations, func(i, j int) bool {
```

Comment:
does this mean we can bind this executor to some other node than `node`? Wouldn't that break the guarantee that the extender will always return a node that exists in the list of given candidate nodes?

Reply:
`ReserveForExecutor` will only bind the executor to the passed node. 
The sorting in this case would prefer choosing the **reservation name** to pick for the (executor, node) pair as a reservation that was previously attached to the same node rather than replacing another node in a separate executor.
Note that I don't think this is important for correctness, I just found that it would be better to do this since the alternative means the nodes in the resource reservation object we first create would frequently be overriden when it is not needed. Happy to remove if you think the added complexity does not bring value.

Reply:
Changed this to 2 separate methods each without a sorting to make the logic explicit rather than implicit.